src/Error/ErrorTrap.php

```diff
@@ -0,0 +1,254 @@
+<?php
+declare(strict_types=1);
+
+namespace Cake\Error;
+
+use Cake\Core\InstanceConfigTrait;
+use Cake\Error\Renderer\ConsoleRenderer;
+use Cake\Error\Renderer\HtmlRenderer;
+use Closure;
+use Exception;
+use InvalidArgumentException;
+use LogicException;
+
+/**
+ * Entry point to CakePHP's error handling.
+ *
+ * Using the `register()` method you can attach an ErrorTrap
+ * to PHP's default error handler. When errors are trapped
+ * they are 'rendered' using the defined renderers and logged
+ * if logging is enabled.
+ */
+class ErrorTrap
+{
+    use InstanceConfigTrait;
+
+    /**
+     * See the `Error` key in you `config/app.php`
+     * for details on the keys and their values.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'errorLevel' => E_ALL,
+        'ignoredDeprecationPaths' => [],
+        'log' => true,
+        'logger' => ErrorLogger::class,
+        'errorRenderer' => null,
+        'extraFatalErrorMemory' => 4 * 1024,
+        'trace' => false,
+    ];
+
+    /**
+     * A list of handling callbacks.
+     *
+     * Callbacks are invoked for each error that is handled.
+     * Callbacks are invoked in the order they are attached.
+     *
+     * @var array<\Closure>
+     */
+    protected $callbacks = [];
+
+    /**
+     * Whether or not this error trap has been registered
+     * as the default error handler. Even when true, this
+     * error handler may no longer be the active error handler.
+     *
+     * @var bool
+     */
+    protected $registered = false;
+
+    /**
+     * Constructor
+     *
+     * @param array $options An options array. See $_defaultConfig.
+     */
+    public function __construct(array $options = [])
+    {
+        $this->setConfig($options);
+        $this->setErrorRenderer($this->chooseErrorRenderer());
+        $this->setLogger($this->getConfig('logger'));
+        $this->setLevel($this->getConfig('errorLevel'));
+    }
+
+    /**
+     * Choose an error renderer based on config or the SAPI
+     *
+     * @return class-string<\Cake\Error\ErrorRendererInterface>
+     */
+    protected function chooseErrorRenderer(): string
+    {
+        $config = $this->getConfig('errorRenderer');
+        if ($config !== null) {
+            return $config;
+        }
+
+        /** @var class-string<\Cake\Error\ErrorRendererInterface> */
+        return PHP_SAPI === 'cli' ? ConsoleRenderer::class : HtmlRenderer::class;
+    }
+
+    /**
+     * Attach this ErrorTrap to PHP's default error handler.
+     *
+     * This will replace the existing error handler, and the
+     * previous error handler will be discarded.
+     *
+     * @return void
+     */
+    public function register(): void
+    {
+        $this->registered = true;
+
+        $level = $this->_config['errorLevel'] ?? -1;
+        error_reporting($level);
```

Comment:
Was this always coupled with registering? We should document this side-effect if so.